File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def __init__(self):
        self.ans = []
    def help(self,index,cand,target,temp):
        if target<0 or index >= len(cand):
            return
        if target == 0:
            
            self.ans.append(temp[:])
            return 
        temp.append(cand[index])
        self.help(index,cand,target - cand[index],temp)
        temp.pop()
        self.help(index+1,cand,target,temp)

    def combinationSum(self, candidates, target):
        self.help(0,candidates,target,[])
        return self.ans
        

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void help(vector<int>cand,int target,int sum,int pos,vector<int>&temp)
    {
        if(sum == target)
        {
            ans.push_back(temp);
            return;
        }
        if(pos>=cand.size() || sum>target)
        {
            return;
        }
        temp.push_back(cand[pos]);
        help(cand,target,sum + cand[pos],pos,temp);
        temp.pop_back();
        help(cand,target,sum,pos+1,temp);

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) 
    {
        vector<int>temp;
        help(candidates,target,0,0,temp);    
        return ans;
    }
};